### PR TITLE
[Torch] Canonicalize pool ops with single int tuple params.

### PR DESCRIPTION
**Pull request by sahas3**
Fixes https://github.com/llvm/torch-mlir/issues/3885 by repeating the single int to match with expected spatial dims.

### CODE REVIEW

**Review by sjarus — APPROVED** [review]

File: include/torch-mlir/Dialect/Torch/IR/GeneratedTorchOps.td [hunk 1]
```diff
@@ -7744,6 +7744,7 @@ def Torch_AtenMaxPool2dOp : Torch_Op<"aten.max_pool2d", [
       printDefaultTorchOp(printer, *this, 6, 1);
     }
   }];
+  let hasCanonicalizer = 1;
```
Comment (vivekkhandelwal1) [inline include/torch-mlir/Dialect/Torch/IR/GeneratedTorchOps.td:7747]: Instead of adding this directly here, you should modify the op registry in `torch_ods_gen.py`.

File: lib/Dialect/Torch/IR/TorchOps.cpp [hunk 2]
```diff
@@ -5601,6 +5601,184 @@ void Aten_AdaptiveAvgPool2dOp::getCanonicalizationPatterns(
...
+
+    auto numSpatialDims = 2;
+    if constexpr (std::is_same<MaxPoolOpT, AtenMaxPool3dOp>())
+      numSpatialDims = 3;
+
+    // Attempt to expand params if necessary.
+    if (failed(expandPoolParams(op, numSpatialDims, rewriter, kernel, stride,
+                                pad, dilations)))
+      return rewriter.notifyMatchFailure(op,
+                                         "Failed to expand params for pooling");
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/IR/TorchOps.cpp]: Instead of pooling just say Max/Avg Pool, since you have 2 different methods.

File: lib/Dialect/Torch/IR/TorchOps.cpp [hunk 3]
```diff
@@ -5601,6 +5601,184 @@ void Aten_AdaptiveAvgPool2dOp::getCanonicalizationPatterns(
...
+          "Expected one of kernel/stride/padding/dilation to be singleton.");
+    }
+
+    expand(dilationInts, numSpatialDims);
+
+  } else if (kernelSizeInts.size() != 1 && paddingInts.size() != 1 &&
+             strideInts.size() != 1) {
+    return rewriter.notifyMatchFailure(
+        op, "Expected one of kernel/stride/padding to be singleton.");
+  }
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/IR/TorchOps.cpp:5773]: Instead of this, you can also do:
```
bool isMaxPool = false;
if constexpr(....) {
    // const dilation check
    isMaxPool = true;
}

// Singleton check for dilation based on the maxpool flag.
// Singleton check for rest of the values.
```

**Review by vivekkhandelwal1 — DISMISSED** [review]

File: lib/Dialect/Torch/IR/TorchOps.cpp (hunk 3 above)
Comment (sahas3) [reply lib/Dialect/Torch/IR/TorchOps.cpp:5773]: I think the singleton check for dilation cannot be decoupled from the singleton check for the other values for `MaxPool` op. For `MaxPool` as long as one of the params of kernel/stride/padding/dilation is singleton we can canonicalize it. For `AvgPool` similarly we have to check for either of kernel/stride/padding to be non-singleton.

So the code will look like:

```
bool isMaxPool = false;
if constexpr(....) { 
    // const dilation check 
    isMaxPool = true; 
}

if (isMaxPool) {
    // check for one of kernel/stride/padding/dilation to be singleton
} else {
   // one of kernel/stride/padding to be singleton
}

if (isMaxPool) {
    expandDilation
} 

// expand other params
```

Is that your suggestion?

File: include/torch-mlir/Dialect/Torch/IR/GeneratedTorchOps.td (hunk 1 above)
Comment (sahas3) [reply include/torch-mlir/Dialect/Torch/IR/GeneratedTorchOps.td:7747]: Done.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py [hunk 4]
```diff
@@ -1810,6 +1861,31 @@ def AvgPool3dCountIncludePadFalseWithoutPadding_basic(module, tu: TestUtils):
     module.forward(tu.rand(3, 3, 12, 12, 12, low=-1))
 
 
+class AvgPool3dSingleIntTupleParamsModule(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.apd = torch.nn.AvgPool3d(
+            kernel_size=(6, 6, 6),
+            stride=(2,),
```
Comment (ivangarcia44) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py:1918]: If the focus of the test is a single value for the stride, then the name of the test could reflect this.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py [hunk 5]
```diff
@@ -722,6 +747,32 @@ def MaxPool3dCeilModeTrueModule_basic(module, tu: TestUtils):
...
 
 
+class MaxPool3dSingleIntTupleParamsModule(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.mpd = torch.nn.MaxPool3d(
+            kernel_size=(6, 6, 6),
+            stride=(2, 2, 2),
+            padding=(1, 1, 1),
+            dilation=(2,),
```
Comment (ivangarcia44) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py:806]: If the goal of the test is the single value dilation the name of the test should reflect it.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py [hunk 6]
```diff
@@ -550,6 +550,31 @@ def MaxPool2dCeilModeFullDimIndivisibleByStrideModule_basic(module, tu: TestUtil
     module.forward(tu.rand(1, 1, 75, 75, low=-1))
 
 
+class MaxPool2dSingleIntTupleParamsModule(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.mpd = torch.nn.MaxPool2d(
+            kernel_size=(6,),
```
Comment (ivangarcia44) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py:606]: If the goal of the test is the single value kernel size the name of the test should reflect it.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py [hunk 7]
```diff
@@ -1810,6 +1861,31 @@ def AvgPool3dCountIncludePadFalseWithoutPadding_basic(module, tu: TestUtils):
     module.forward(tu.rand(3, 3, 12, 12, 12, low=-1))
 
 
+class AvgPool3dSingleIntTupleParamsModule(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.apd = torch.nn.AvgPool3d(
+            kernel_size=(6, 6, 6),
+            stride=(2,),
+            padding=(1, 1, 1),
```
Comment (ivangarcia44) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py:1919]: What about a test for single value for padding?

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py [hunk 8]
```diff
@@ -1810,6 +1861,31 @@ def AvgPool3dCountIncludePadFalseWithoutPadding_basic(module, tu: TestUtils):
     module.forward(tu.rand(3, 3, 12, 12, 12, low=-1))
 
 
+class AvgPool3dSingleIntTupleParamsModule(torch.nn.Module):
```
Comment (ivangarcia44) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py]: The kernel size and dilation are tested with max pooling, and the stride with average pooling. If they share the same path this is fine. But I saw the max pooling taking a separate path for expansion. If the paths are not shared, it might be good to have test for max pooling and average pooling for all the parameters that could be expanded. If adding these tests do not increate path coverage in the source code, you can ignore this message.

Comment (ivangarcia44) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py]: On a second look, the dilation path is only for max pooling, average pooling does not have it. The expansion for the other 3 parameters is shared. You can ignore this message.

File: lib/Dialect/Torch/IR/TorchOps.cpp (hunk 3 above)
Comment (sahas3) [reply lib/Dialect/Torch/IR/TorchOps.cpp:5773]: Hi @vivekkhandelwal1 can you please clarify if I understood your suggestion as captured in my previous comment? If yes, is it to make the code easier to read?

Comment (sahas3) [reply lib/Dialect/Torch/IR/TorchOps.cpp:5773]: Hi @vivekkhandelwal1 any further thoughts on this?

**Comment by sahas3** [comment]
Hi @vivekkhandelwal1 do you have any further comments on this? Since this has been open for a while, I'll go ahead with merging the PR if I don't hear back from you by this week. I'll be happy to address any further comments afterwards as well. Thanks!
